main/ZactuatorFASTLED.ino

```diff
@@ -0,0 +1,293 @@
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "User_config.h"
+
+#ifdef ZactuatorFASTLED
+
+#ifdef ESP8266
+#include <FastLED.h>
+#else
+#include <FastLED.h>
+#endif
+#include <ArduinoJson.h>
+enum LEDState
+{
+  OFF,
+  FIRE,
+  GENERAL
+};
+LEDState currentLEDState;
+long lastUpdate = 0;
+long currentUpdate = 0;
+CRGB leds[FASTLED_NUM_LEDS];
+CRGB ledColorBlink[FASTLED_NUM_LEDS];
+StaticJsonBuffer<200> jsonBuffer;
+bool blinkLED[FASTLED_NUM_LEDS];
+const long blinkInterval = 300;
+const long fireUpdate = 10;
+unsigned long previousMillis = 0;
+CRGBPalette16 gPal;
+
+void setupFASTLED()
+{
+  trc(F("FASTLED_DATA_PIN "));
+  trc(String(FASTLED_DATA_PIN));
+  trc(F("FASTLED_NUM_LEDS "));
+  trc(String(FASTLED_NUM_LEDS));
+  trc(F("ZactuatorFASTLED setup done "));
+  FastLED.addLeds<FASTLED_TYPE, FASTLED_DATA_PIN>(leds, FASTLED_NUM_LEDS);
+}
+
+//returns the current step of the animation
+int animation_step(int duration, int steps)
+{
+  int currentStep = ((currentUpdate % duration) / ((float)duration)) * steps;
+  return currentStep;
+}
+
+//returns the number of steps since the last update
+int animation_step_count(int duration, int steps)
+{
+  long lastAnimationNumber = lastUpdate / duration;
+  long currentAnimationNumber = currentUpdate / duration;
+  int lastStep = ((lastUpdate % duration) / ((float)duration)) * steps;
+  int currentStep = ((currentUpdate % duration) / ((float)duration)) * steps;
+
+  return currentStep - lastStep + (currentAnimationNumber - lastAnimationNumber) * steps;
+}
+
+void FASTLEDLoop()
+{
+
+  lastUpdate = currentUpdate;
+  currentUpdate = millis();
+
+  if (currentLEDState == GENERAL)
+  {
+
+    for (int i = 0; i < FASTLED_NUM_LEDS; i++)
+    {
+
+      int count = animation_step_count(blinkInterval, 2);
+      if (count > 0)
+      {
+        int step = animation_step(blinkInterval, 2);
+
+        if (blinkLED[i])
+        {
+
+          if (step == 0)
+          {
+            leds[i] = ledColorBlink[i];
+          }
+          else
+          {
+            ledColorBlink[i] = leds[i];
+            leds[i] = CRGB::Black;
+          }
+        }
+      }
+    }
+  }
+  else if (currentLEDState == FIRE)
+  {
+    int count = animation_step_count(fireUpdate, 2);
+    if (count > 0)
+    {
+      //int step = animation_step(fireUpdate, 2);
+      //random16_add_entropy( random(0)); //random() don't exists in ESP framework - workaround?
+
+      Fire2012WithPalette();
+    }
+  }
+  FastLED.show();
+}
+
+boolean FASTLEDtoMQTT()
+{
+  return false;
+}
+void MQTTtoFASTLEDJSON(char *topicOri, JsonObject&  jsonData)
+{
+  trc(F("MQTTtoFASTLEDJSON: "));
+  currentLEDState = GENERAL;
+  String topic = topicOri;
+  
+  trc(topic);
+  //number = (long)strtol(&datacallback[1], NULL, 16);
+
+  
+  
+  if (topic == subjectMQTTtoFASTLEDsetled)
+  {
+    
+    
+    
+   
```

Comment:
remove the empty lines. One empty line is ok for readability.

Reply:
I updated the code